[Discover] Redesign for the grid, panels and sidebar v1

jughosta opened the pull request:
## Summary

### Part 1

- Resolves https://github.com/elastic/kibana/issues/164287
- Closes https://github.com/elastic/kibana/issues/146339
- Previously separate PR https://github.com/elastic/kibana/pull/164187

Changes:
- ~~swaps checkbox and row selection~~
- removes vertical borders
- adds rows highlight
- increases cell padding
- adds row stripes
- updates header background
- removes grey background from field name and makes it bolder (part of https://github.com/elastic/kibana/issues/164634)
- updates Surrounding Documents side paddings

### Part 2

- Resolves https://github.com/elastic/kibana/issues/164661
- Previously separate PR https://github.com/elastic/kibana/pull/165687

Changes:
- removes background from panels, tabs and sidebar
- updates "Add a field" button style
- removes shadow from field list items
- makes field search compact

### Part 3

- Resolves https://github.com/elastic/kibana/issues/164662

Changes:
- wraps "Add a field" button in its own container with a top border
- ~~adds a drag handle to sidebar items~~
- ~~adds new Show/Hide buttons to toggle sidebar~~ moves sidebar toggle button from discover plugin to unified field list
- reduces spaces between sidebar items from 4px to 2px
- reduces padding on Single Document page
- removes border above grid tabs

<img width="600" alt="Screenshot 2023-09-07 at 14 39 48" src="https://github.com/elastic/kibana/assets/1415710/976db247-fd70-4c9b-8634-552ece45b522">


Please note that "auto" row height is in a separate PR which is also ready for review https://github.com/elastic/kibana/pull/164218

elasticmachine commented:
Pinging @elastic/kibana-data-discovery (Team:DataDiscovery)

jughosta commented:
Still waiting for CI to see if some tests fail but the functionality is ready for preview and review.

andreadelrio reviewed (changes requested):
I see that the new icons are not working properly in dark mode. My recommendation is that we don't rush the addition of these icons and wait until they're added to EUI. This will save us both development and review time. We're tracking the addition of these icons to EUI here https://github.com/elastic/eui/issues/7162 https://github.com/elastic/eui/issues/7141

davismcphee reviewed (commented):
Awesome work! The changes are looking great and I haven't noticed any UI issues when testing in both light and dark mode. Looks like there are a few tests left to address, but overall looking really good.

Since we updated the surrounding docs padding, I wonder if we should do the same to the single doc view? I've always thought it seemed too large at 24px and looks much nicer at 16px. WDYT?

24px:
<img width="1186" alt="padding_before" src="https://github.com/elastic/kibana/assets/25592674/a7548c46-19e4-47d8-aab5-62f8a7cab762">

16px:
<img width="1186" alt="padding_after" src="https://github.com/elastic/kibana/assets/25592674/a57e07ee-271d-4e92-b1e0-d8dc82973323">

Otherwise my only feedback is that since we removed most of the borders elsewhere, it really draws attention to the stack of horizontal borders at the top of the data grid. It looks a little strange IMO and maybe we could get rid of some of them, or maybe it's fine for now until we do something about the field statistics tab:
<img width="1278" alt="horiz_borders" src="https://github.com/elastic/kibana/assets/25592674/a25e3a6d-fa78-497c-8f7b-11041c9ccc9e">

>I see that the new icons are not working properly in dark mode. My recommendation is that we don't rush the addition of these icons and wait until they're added to EUI.

@andreadelrio Thanks for the feedback. Should we include the new field list changes other than the icons and just use the old icons for the collapse button and no drag icon for the field list items, or should we revert these changes entirely until the icons are available?

andreadelrio commented:
> Since we updated the surrounding docs padding, I wonder if we should do the same to the single doc view? I've always thought it seemed too large at 24px and looks much nicer at 16px. WDYT?

I like it! 

> Otherwise my only feedback is that since we removed most of the borders elsewhere, it really draws attention to the stack of horizontal borders at the top of the data grid. It looks a little strange IMO and maybe we could get rid of some of them, or maybe it's fine for now until we do something about the field statistics tab: <img alt="horiz_borders" width="1278" src="https://user-images.githubusercontent.com/25592674/266181405-a25e3a6d-fa78-497c-8f7b-11041c9ccc9e.png">

Agreed. I think we can remove the horizontal line above the tabs.

<img width="880" alt="image" src="https://github.com/elastic/kibana/assets/4016496/d1ecc7a5-ada8-4f7c-b4c1-43aaa459f502">

> @andreadelrio Thanks for the feedback. Should we include the new field list changes other than the icons and just use the old icons for the collapse button and no drag icon for the field list items, or should we revert these changes entirely until the icons are available?

I say we go ahead with the rest of the field list changes other than the icons. Continue using the old icon for collapsing the field list (for now) and no drag icon for field list items (for now). Michael is working on adding the icons to EUI.

jughosta commented on src/plugins/discover/public/application/main/components/layout/discover_layout.tsx:
```diff
@@ -284,31 +296,21 @@ export function DiscoverLayout({ stateContainer }: DiscoverLayoutProps) {
-                  })}
-                />
-              </div>
-            </EuiFlexItem>
+            <EuiFlexItem
+              grow={false}
+              css={css`
+                border-right: ${euiTheme.border.thin};
+              `}
+            />
           </EuiHideFor>
```
@davismcphee I'm not sure about it. On Lens page for example there is no right border for the sidebar at the moment.

@andreadelrio Should we include the right border inside the unified field list or it should only stay outside for Discover view?

jughosta commented on src/plugins/discover/public/application/main/components/sidebar/discover_sidebar_responsive.tsx:
```diff
@@ -380,7 +386,7 @@ export function DiscoverSidebarResponsive(props: DiscoverSidebarResponsiveProps)
       variant={fieldListVariant}
       getCreationOptions={getCreationOptions}
-      isSidebarCollapsed={props.isClosed}
+      isSidebarCollapsed={isSidebarCollapsed}
```
Updated via https://github.com/elastic/kibana/pull/165866/commits/12f56beb6907d2696d0dd9fe30d25dff0510cbbb

kertal commented on packages/kbn-unified-data-table/src/components/data_table.scss:
```diff
@@ -118,7 +109,16 @@
   @include euiTextTruncate;
 }
 
+.unifiedDataTable__descriptionListTitle {
+  margin-inline: 0 0;
+  padding-inline: 0;
+  background: transparent;
+  font-weight: $euiFontWeightBold;
+}
+
 .unifiedDataTable__descriptionListDescription {
+  margin-inline: $euiSizeS $euiSizeS;
+  padding-inline: 0;
```
worth to note that these changes will also resolve #146339  🥳

jughosta commented:
@davismcphee @andreadelrio Thanks for the feedback! I updated the PR.

@davismcphee What could trigger `discover/group1 discover histogram should modify the time range when the histogram is brushed` test failures?

MichaelMarcialis commented:
> I see that the new icons are not working properly in dark mode. My recommendation is that we don't rush the addition of these icons and wait until they're added to EUI. This will save us both development and review time. We're tracking the addition of these icons to EUI here [elastic/eui#7162](https://github.com/elastic/eui/issues/7162) [elastic/eui#7141](https://github.com/elastic/eui/issues/7141)

@andreadelrio, @jughosta: FYI, I've opened a [PR for the new icons to be added to EUI](https://github.com/elastic/eui/pull/7168). It should hopefully be merged in shortly and part of the next EUI update in Kibana.

davismcphee reviewed (commented):
Thanks for the latest changes! This is looking very close to being good to go on my end, just a few more comments and suggestions.

It looks like the "Add a field" button is no longer working in mobile:
![add_field](https://github.com/elastic/kibana/assets/25592674/e6a28672-ed78-4cb9-adfd-9606371e42f5)

I noticed there is no longer a border shown beneath the tabs when switching to Field statistics which makes the tab switch a little jarring. I think it would be good to add this back in for now:
<img width="1816" alt="field_stats_hr" src="https://github.com/elastic/kibana/assets/25592674/2ffed2eb-7ce2-48bd-a387-aa1c919b0545">

Since removing the drag handle from the field list items, it looks like there is no longer left padding around the icon when dragging, which looks a little odd:
<img width="1197" alt="drag_padding" src="https://github.com/elastic/kibana/assets/25592674/13bbc749-355d-4d48-b663-5a35d6d54f53">

The drop shadows on the compressed field list items are gone, but based on the examples page it looks like they still have white backgrounds. It's not noticeable in Discover since the page background is white too, but it might be worth making the compressed field list item backgrounds transparent regardless for future consumers:
<img width="1816" alt="example_field_list" src="https://github.com/elastic/kibana/assets/25592674/8c268530-6ed9-4a4d-9fe5-d6568b6dfd7e">

The change to using a compressed input for the field list filter makes it so the input no longer lines up with the top controls in Lens. I'll leave it up to @andreadelrio to decide if we're ok with that or if we should only use a compressed input when using compressed field list items:
<img width="1816" alt="lens_field_list" src="https://github.com/elastic/kibana/assets/25592674/ea337f52-6a6d-4a4e-bdf1-0b3675c9f7a4">

And lastly, I think it looks much better with the border removed above the tabs 👍 My only complaint now is that it feels like there's some excess whitespace between the chart, tabs, and resize button. It's a minor concern, but I think it would look better to compress this space a bit and squeeze a few more pixels out of the chart where it counts most. I put together [a tiny PR](https://github.com/jughosta/kibana/pull/5) including these changes if you think we should add them:

Current:
<img width="1816" alt="resize_handle_current" src="https://github.com/elastic/kibana/assets/25592674/37a5a9da-3440-4454-8909-c8137110cdd0">

Compressed:
<img width="1816" alt="resize_handle_compressed" src="https://github.com/elastic/kibana/assets/25592674/a0f33703-4013-4a85-9ed7-b9ac218379dc">

andreadelrio commented:
> The drop shadows on the compressed field list items are gone, but based on the examples page it looks like they still have white backgrounds. It's not noticeable in Discover since the page background is white too, but it might be worth making the compressed field list item backgrounds transparent regardless for future consumers: <img alt="example_field_list" 
> 
+1 
> The change to using a compressed input for the field list filter makes it so the input no longer lines up with the top controls in Lens. I'll leave it up to @andreadelrio to decide if we're ok with that or if we should only use a compressed input when using compressed field list items: <img alt="lens_field_list" width="1816" src="https://user-images.githubusercontent.com/25592674/266429937-ea337f52-6a6d-4a4e-bdf1-0b3675c9f7a4.png">

I think we should only use compressed input when using the compressed list items. Also in the screenshot above the compressed field list filter next to the regular-sized toolbar on the right looks off. If Lens wanted to adopt the compressed field list it would make sense to use the compressed version of the Shared UX buttons in its toolbar too but I defer to @MichaelMarcialis for all Lens things.

I was going to suggest disabling dragging on the selected fields until I remembered we left it like that because users can drop into the query bar (I still don't know how much of a use case this actually is). In any case, I noticed that this doesn't seem to be working in ES|QL.
![Screen Recording 2023-09-07 at 1 56 02 PM](https://github.com/elastic/kibana/assets/4016496/6a1621a9-2274-4e14-8bff-692ecaa7c4cf)

jughosta commented on packages/kbn-unified-field-list/src/containers/unified_field_list_sidebar/field_list_sidebar.tsx:
```diff
@@ -95,6 +97,11 @@ interface UnifiedFieldListSidebarInternalProps {
+  /**
+   * What button style type to use
+   */
+  buttonAddFieldVariant: ButtonAddFieldVariant;
```
@davismcphee I kept it flexible as in some context "toolbar" style with a border might not fit (as on Examples page):

<img width="430" alt="Screenshot 2023-09-08 at 12 10 03" src="https://github.com/elastic/kibana/assets/1415710/5aedd562-5eaa-4e44-9e1a-cdd4f60f6b6d">

In this case, the previous style could be used:

<img width="363" alt="Screenshot 2023-09-08 at 12 10 27" src="https://github.com/elastic/kibana/assets/1415710/a015490f-44ad-4658-86d7-d94721736d59">

jughosta commented:
@davismcphee Thanks for catching these! I updated the branch.

andreadelrio commented:
@jughosta I noticed that there is something off when `Row height` is set to `Single`.
<img width="341" alt="image" src="https://github.com/elastic/kibana/assets/4016496/7ecfa9b3-7de9-43c7-9293-13fc75463516">

davismcphee reviewed (commented):
Thanks for the latest changes! They look good for the most part, but I noticed that because we added the padding back to the list item when dragging, it meant we needed to add it to the regular list item too, which doesn't match the designs. It also seems like the white background will reappear when a list item is active, which looks a bit strange:
<img width="1196" alt="list_item" src="https://github.com/elastic/kibana/assets/25592674/d9ffb19c-f5f6-4c4c-a601-66019709938c">

I think I found a fix for it though where we can add a custom class to style the list item while dragging without having to change the regular list items too. I did some testing across browsers and it seems to work consistently. I opened up a PR for it here, let me know what you think and if it makes sense to merge: https://github.com/jughosta/kibana/pull/6.

I also noticed that when viewing the documents column in single line mode, the line below is partially visible:
<img width="1266" alt="single_line" src="https://github.com/elastic/kibana/assets/25592674/3110822f-9911-42b3-b544-b9ab036affda">

jughosta commented:
@andreadelrio 

>  I noticed that there is something off when Row height is set to Single.

@davismcphee 

> I also noticed that when viewing the documents column in single line mode, the line below is partially visible

Addressed these issues via https://github.com/elastic/kibana/pull/165866/commits/a93943c4ec4d09b736e31422ce0b664c58400fda

jughosta commented on packages/kbn-dom-drag-drop/src/drag_drop.tsx:
```diff
@@ -305,6 +310,18 @@ const DragInner = memo(function DragInner({
       // so we know we have DraggableProps if we reach this code.
       if (e && 'dataTransfer' in e) {
         e.dataTransfer.setData('text', value.humanData.label);
+
+        // Apply an optional class to the element being dragged so the ghost
+        // can be styled. We must add it to the actual element for a single
+        // frame before removing it so the ghost picks up the styling.
+        const current = e.currentTarget;
+
+        if (dragClassName && !current.classList.contains(dragClassName)) {
+          current.classList.add(dragClassName);
+          requestAnimationFrame(() => {
+            current.classList.remove(dragClassName);
+          });
+        }
```
@davismcphee helped here to improve the view of a dragging item on Discover https://github.com/jughosta/kibana/pull/6

andreadelrio reviewed (approved):
Design changes for v1 LGTM! Nice job @jughosta.

davismcphee reviewed (approved):
Thanks for the fixes! I did a final round of testing and confirmed the latest issues were fixed, and I didn't notice anything else. Awesome work on this, it's such a big improvement to Discover's UI, LGTM 👍

stratoula reviewed (approved):
This is indeed awesome Julia! LGTM!

qd-danh commented:
not liking the look of the data grid without vertical lines. Now the data columns just seem to fall off into the empty space.  Is this a global change and will remain going forward? 

Wish this was configurable based on a setting somewhere.

before
![image](https://github.com/elastic/kibana/assets/85181651/8ca3db6e-9bb6-4e2e-8318-6955e0ae8489)

after
![image](https://github.com/elastic/kibana/assets/85181651/6e6e5c43-86a7-4f70-b252-fcd683650907)
